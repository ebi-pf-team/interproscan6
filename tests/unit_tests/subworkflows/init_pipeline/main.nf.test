nextflow_workflow {

    name "Test suborkflow INIT_PIPELINE"
    script "subworkflows/init/main.nf"
    workflow "INIT_PIPELINE"

    stage {
        symlink "tests/data/sequences/test.faa"
        symlink "conf/applications.config"
    }

    test("Should run without failures") {
        when {
            params {
                apps_config = [
                    pfam: [
                        name: "Pfam",
                        dir: "pfam",
                        hmm: "pfam_a.hmm",
                        seed: "pfam_a.seed",
                        clan: "pfam_clans",
                        dat: "pfam_a.dat",
                        has_data: true
                    ],
                    ncbifam: [
                        name: "NCBIFAM",
                        dir: "ncbifam",
                        hmm: "ncbifam.hmm",
                        has_data: true
                    ]
                ]
            }
            workflow {
                """
                input[0] = "meta/tests/data/sequences/test.faa"     // input
                input[1] = "pfam,ncbifam"                           // applictaions
                input[2] = params.apps_config                       // apps_config
                input[3] = "data"                                   // datadir
                input[4] = "json,tsv,xml"                           // formats
                input[5] = "results"                                // outdir
                input[6] = null                                     // outprefix  
                input[7] = false                                    // no_matches_api
                input[8] = "latest"                                 // interpro_version
                input[9] = false                                    // skip_intepro
                input[10] = null                                    // skip_applications
                input[11] = false                                   // goterms
                input[12] = false                                   // pathways
                """
            }
        }
        then {
            assert workflow.success
            assert snapshot(workflow.out).match()
        }
    }

    test("Should fail: No input") {
        when {
            params {
                apps_config = [
                    pfam: [
                        name: "Pfam",
                        dir: "pfam",
                        hmm: "pfam_a.hmm",
                        seed: "pfam_a.seed",
                        clan: "pfam_clans",
                        dat: "pfam_a.dat",
                        has_data: true
                    ],
                    ncbifam: [
                        name: "NCBIFAM",
                        dir: "ncbifam",
                        hmm: "ncbifam.hmm",
                        has_data: true
                    ]
                ]
            }
            workflow {
                """
                input[0] = "meta/tests/data/non-existent-faa.faa"        // input
                input[1] = "pfam,ncbifam"                    // applictaions
                input[2] = params.apps_config                // apps_config
                input[3] = false                             // download
                input[4] = false                             // offline
                input[5] = "data"                            // datadir
                input[6] = "json,tsv,xml"                    // formats
                input[7] = "outdir"                          // outdir
                input[8] = "latest"                          // interpro_version
                input[9] = false                             // skip_interpro
                input[10] = false                            // goterms
                input[11] = false                            // pathways
                """
            }
        }

        then {
            assert workflow.failed
            assert workflow.exitStatus == 1
        }
    }

    test("Should fail: Invalid applications") {
        when {
            params {
                apps_config = [
                    pfam: [
                        name: "Pfam",
                        dir: "pfam",
                        hmm: "pfam_a.hmm",
                        seed: "pfam_a.seed",
                        clan: "pfam_clans",
                        dat: "pfam_a.dat",
                        has_data: true
                    ],
                    ncbifam: [
                        name: "NCBIFAM",
                        dir: "ncbifam",
                        hmm: "ncbifam.hmm",
                        has_data: true
                    ]
                ]
            }
            workflow {
                """
                input[0] = "meta/tests/data/sequences/test.faa"        // input
                input[1] = "pfam,ncbifam,fake_application"                    // applictaions
                input[2] = params.apps_config                // apps_config
                input[3] = false                             // download
                input[4] = false                             // offline
                input[5] = "data"                            // datadir
                input[6] = "json,tsv,xml"                    // formats
                input[7] = "outdir"                          // outdir
                input[8] = "latest"                          // interpro_version
                input[9] = false                             // skip_interpro
                input[10] = false                            // goterms
                input[11] = false                            // pathways
                """
            }
        }
        then {
            assert workflow.failed
            assert workflow.exitStatus == 1
        }
    }

    test("Should fail: Download and Offline enabled") {
        when {
            params {
                apps_config = [
                    pfam: [
                        name: "Pfam",
                        dir: "pfam",
                        hmm: "pfam_a.hmm",
                        seed: "pfam_a.seed",
                        clan: "pfam_clans",
                        dat: "pfam_a.dat",
                        has_data: true
                    ],
                    ncbifam: [
                        name: "NCBIFAM",
                        dir: "ncbifam",
                        hmm: "ncbifam.hmm",
                        has_data: true
                    ]
                ]
            }
            workflow {
                """
                input[0] = "meta/tests/data/sequences/test.faa"        // input
                input[1] = "pfam,ncbifam"                    // applictaions
                input[2] = params.apps_config                // apps_config
                input[3] = true                              // download
                input[4] = true                              // offline
                input[5] = "data"                            // datadir
                input[6] = "json,tsv,xml"                    // formats
                input[7] = "outdir"                          // outdir
                input[8] = "latest"                          // interpro_version
                input[9] = false                             // skip_interpro
                input[10] = false                            // goterms
                input[11] = false                            // pathways
                """
            }
        }
        then {
            assert workflow.failed
            assert workflow.exitStatus == 1
        }
    }

    test("Should fail: Skip-interpro, goterms and pathways enabled") {
        when {
            params {
                apps_config = [
                    pfam: [
                        name: "Pfam",
                        dir: "pfam",
                        hmm: "pfam_a.hmm",
                        seed: "pfam_a.seed",
                        clan: "pfam_clans",
                        dat: "pfam_a.dat",
                        has_data: true
                    ],
                    ncbifam: [
                        name: "NCBIFAM",
                        dir: "ncbifam",
                        hmm: "ncbifam.hmm",
                        has_data: true
                    ]
                ]
            }
            workflow {
                """
                input[0] = "meta/tests/data/sequences/test.faa"        // input
                input[1] = "pfam,ncbifam"                    // applictaions
                input[2] = params.apps_config                // apps_config
                input[3] = false                             // download
                input[4] = false                             // offline
                input[5] = "data"                            // datadir
                input[6] = "json,tsv,xml"                    // formats
                input[7] = "outdir"                          // outdir
                input[8] = "latest"                          // interpro_version
                input[9] = true                              // skip_interpro
                input[10] = true                             // goterms
                input[11] = true                            // pathways
                """
            }
        }
        then {
            assert workflow.failed
            assert workflow.exitStatus == 1
        }
    }

    test("Should fail: Invalid formats") {
        when {
            params {
                apps_config = [
                    pfam: [
                        name: "Pfam",
                        dir: "pfam",
                        hmm: "pfam_a.hmm",
                        seed: "pfam_a.seed",
                        clan: "pfam_clans",
                        dat: "pfam_a.dat",
                        has_data: true
                    ],
                    ncbifam: [
                        name: "NCBIFAM",
                        dir: "ncbifam",
                        hmm: "ncbifam.hmm",
                        has_data: true
                    ]
                ]
            }
            workflow {
                """
                input[0] = "meta/tests/data/sequences/test.faa"        // input
                input[1] = "pfam,ncbifam"                    // applictaions
                input[2] = params.apps_config                // apps_config
                input[3] = false                              // download
                input[4] = false                              // offline
                input[5] = "data"                            // datadir
                input[6] = "json,tsv,xml,FAKE"                    // formats
                input[7] = "outdir"                          // outdir
                input[8] = "latest"                          // interpro_version
                input[9] = false                             // skip_interpro
                input[10] = false                            // goterms
                input[11] = false                            // pathways
                """
            }
        }
        then {
            assert workflow.failed
            assert workflow.exitStatus == 1
        }
    }

    test("Should fail: Datadir not provided") {
        when {
            params {
                apps_config = [
                    pfam: [
                        name: "Pfam",
                        dir: "pfam",
                        hmm: "pfam_a.hmm",
                        seed: "pfam_a.seed",
                        clan: "pfam_clans",
                        dat: "pfam_a.dat",
                        has_data: true
                    ],
                    ncbifam: [
                        name: "NCBIFAM",
                        dir: "ncbifam",
                        hmm: "ncbifam.hmm",
                        has_data: true
                    ]
                ]
            }
            workflow {
                """
                input[0] = "meta/tests/data/sequences/test.faa"        // input
                input[1] = "pfam,ncbifam"                    // applictaions
                input[2] = params.apps_config                // apps_config
                input[3] = false                             // download
                input[4] = false                             // offline
                input[5] = null                              // datadir
                input[6] = "json,tsv,xml"                    // formats
                input[7] = "outdir"                          // outdir
                input[8] = "latest"                          // interpro_version
                input[9] = false                             // skip_interpro
                input[10] = false                            // goterms
                input[11] = false                            // pathways
                """
            }
        }
        then {
            assert workflow.failed
            assert workflow.exitStatus == 1
        }
    }
    
    test("Should fail: Invalid datadir") {
        when {
            params {
                apps_config = [
                    pfam: [
                        name: "Pfam",
                        dir: "pfam",
                        hmm: "pfam_a.hmm",
                        seed: "pfam_a.seed",
                        clan: "pfam_clans",
                        dat: "pfam_a.dat",
                        has_data: true
                    ],
                    ncbifam: [
                        name: "NCBIFAM",
                        dir: "ncbifam",
                        hmm: "ncbifam.hmm",
                        has_data: true
                    ]
                ]
            }
            workflow {
                """
                input[0] = "meta/tests/data/sequences/test.faa"        // input
                input[1] = "pfam,ncbifam"                    // applictaions
                input[2] = params.apps_config                // apps_config
                input[3] = false                             // download
                input[4] = false                             // offline
                input[5] = "meta/tests/data/sequences/test.faa"        // datadir
                input[6] = "json,tsv,xml"                    // formats
                input[7] = "outdir"                          // outdir
                input[8] = "latest"                          // interpro_version
                input[9] = false                             // skip_interpro
                input[10] = false                            // goterms
                input[11] = false                            // pathways
                """
            }
        }
        then {
            assert workflow.failed
            assert workflow.exitStatus == 1
        }
    }
    
    test("Should fail: Invalid interpro version") {
        when {
            params {
                apps_config = [
                    pfam: [
                        name: "Pfam",
                        dir: "pfam",
                        hmm: "pfam_a.hmm",
                        seed: "pfam_a.seed",
                        clan: "pfam_clans",
                        dat: "pfam_a.dat",
                        has_data: true
                    ],
                    ncbifam: [
                        name: "NCBIFAM",
                        dir: "ncbifam",
                        hmm: "ncbifam.hmm",
                        has_data: true
                    ]
                ]
            }
            workflow {
                """
                input[0] = "meta/tests/data/sequences/test.faa"        // input
                input[1] = "pfam,ncbifam"                    // applictaions
                input[2] = params.apps_config                // apps_config
                input[3] = false                             // download
                input[4] = false                             // offline
                input[5] = "data"                            // datadir
                input[6] = "json,tsv,xml"                    // formats
                input[7] = "outdir"                          // outdir
                input[8] = "fake-version"                    // interpro_version
                input[9] = false                             // skip_interpro
                input[10] = false                            // goterms
                input[11] = false                            // pathways
                """
            }
        }
        then {
            assert workflow.failed
            assert workflow.exitStatus == 1
        }
    }
    
    test("Should fail: Invalid apps for online mode") {
        when {
            params {
                apps_config = [
                    signalp_euk: [
                        name: "SignalP-Euk",
                        organism: "eukarya",
                        dir: "data",
                        mode: "fast",
                        has_data: false,
                        use_gpu: false
                    ],
                    signalp_prok: [
                        name: "SignalP-Prok",
                        organism: "eukarya",
                        dir: "data",
                        mode: "fast",
                        has_data: false,
                        use_gpu: false
                    ],
                    deeptmhmm: [
                        name: "DeepTMHMM",
                        dir: "data",
                        has_data: false,
                        use_gpu: false
                    ]
                ]
            }
            workflow {
                """
                input[0] = "meta/tests/data/sequences/test.faa"         // input
                input[1] = "signap-prok,signap-euk,deeptmhmm" // applictaions
                input[2] = params.apps_config                 // apps_config
                input[3] = false                              // download
                input[4] = false                              // offline
                input[5] = "data"                             // datadir
                input[6] = "json,tsv,xml"                     // formats
                input[7] = "outdir"                           // outdir
                input[8] = "fake-version"                     // interpro_version
                input[9] = false                              // skip_interpro
                input[10] = false                             // goterms
                input[11] = false                             // pathways
                """
            }
        }
        then {
            assert workflow.failed
            assert workflow.exitStatus == 1
        }
    }
}
