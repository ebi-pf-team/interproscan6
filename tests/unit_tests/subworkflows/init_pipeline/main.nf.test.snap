{
    "Should run without failures": {
        "content": [
            {
                "0": [
                    "test.faa:md5,ca69be62364cac44c8083faace76014e"
                ],
                "1": [
                    [
                        "ncbifam",
                        "pfam"
                    ]
                ],
                "2": [
                    [
                        
                    ]
                ],
                "3": [
                    "/home/mblum/Projects/i6/interproscan6/.nf-test/tests/aba7a0f4f7a5b34d9d32ab5a3878abae/results/test.faa"
                ],
                "4": [
                    [
                        "TSV",
                        "XML",
                        "JSON"
                    ]
                ],
                "5": [
                    "latest"
                ],
                "apps": [
                    [
                        "ncbifam",
                        "pfam"
                    ]
                ],
                "datadir": [
                    [
                        
                    ]
                ],
                "fasta": [
                    "test.faa:md5,ca69be62364cac44c8083faace76014e"
                ],
                "formats": [
                    [
                        "TSV",
                        "XML",
                        "JSON"
                    ]
                ],
                "outprefix": [
                    "/home/mblum/Projects/i6/interproscan6/.nf-test/tests/aba7a0f4f7a5b34d9d32ab5a3878abae/results/test.faa"
                ],
                "version": [
                    "latest"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-06-13T11:44:28.164313"
    }
}